src/Compiler/Symbols/Exprs.fs

```diff
@@ -1264,9 +1264,12 @@ module FSharpExprConvert =
         match x with 
         | TDSwitch(inpExpr, csl, dfltOpt, m) -> 
             let acc = 
-                match dfltOpt with 
-                | Some d -> ConvDecisionTreePrim cenv env dtreeRetTy d 
-                | None -> wfail( "FSharp.Compiler.Service cannot yet return this kind of pattern match", m)
+                match dfltOpt, dtreeRetTy with 
+                | Some d, _ -> ConvDecisionTreePrim cenv env dtreeRetTy d 
+                | None, TType.TType_app(tyconRef = _tyconRef; typeInstantiation = _typeInstantiation; flags = _flags)
+                    when _tyconRef.CompiledName = "int" || _tyconRef.CompiledName = "bool" ->
```

Comment:
I must admit I do not see why we have to special case exactly int and bool (and nothing else)/
What is the reason?

Reply:
I'm trying to limit the change of behaviour to a minimum while still getting rid of the exception for typed union fields.
And well, this is what seemed to be the minimal change to me.

`AugmentWithHashCompare.fs` adds `CompareTo` and `GetHashCode` which have type `int` and it adds `Equals` which has type `bool`.
These members are causing the `dfltOpt` to be `None` if the Union has typed fields, also see the description of this PR for the exact spots in the code.

But yeah, I also admit I'm not super confident that this is the correct way to do it.

Reply:
Mmh, I can't really think of a meaningful risk. So I'm okay with going without the limiting.

Reply:
In that case, I would suggest:

- Do not limit if things should work anyway
- If limiting for the purpose of auto-generated methods (CompareTo,GetHashCode), limit just those explicitely.